File: 抽象工厂模式.cpp
#include <bits/stdc++.h>
using namespace std;
class AbstractApple
{
public:
    virtual void ShowName() = 0;
};
class ChinaApple : public AbstractApple
{
public:
    virtual void ShowName()
    {
        cout << "i'm Chinaapple" << endl;
    }
};
class USAApple : public AbstractApple
{
public:
    virtual void ShowName()
    {
        cout << "i'm USAapple" << endl;
    }
};
class AbstractBanana
{
public:
    virtual void ShowName() = 0;
};
class ChinaBanana : public AbstractBanana
{
public:
    virtual void ShowName()
    {
        cout << "i'm ChinaBanana" << endl;
    }
};
class USABanana : public AbstractBanana
{
public:
    virtual void ShowName()
    {
        cout << "i'm USABanana" << endl;
    }
};
class AbstractFactory
{
public:
    virtual AbstractApple *CreateApple() = 0;
    virtual AbstractBanana *CreateBanan() = 0;
};
class ChinaFactory : public AbstractFactory
{
public:
    virtual AbstractApple *CreateApple()
    {
        return new ChinaApple;
    }
    virtual AbstractBanana *CreateBanan()
    {
        return new ChinaBanana;
    }
};
class USAFactory : public AbstractApple
{
public:
    virtual AbstractApple *CreateApple()
    {
        return new USAApple;
    }
    virtual AbstractBanana *CreateBanan()
    {
        return new USABanana;
    }
};
int main()
{
    AbstractFactory *factory = new ChinaFactory;
    AbstractApple *fruit = factory->CreateApple();
    fruit->ShowName();
    delete fruit;
    fruit = nullptr;
    delete factory;
    factory = nullptr;
    return 0;
}

File: 抽象工厂模式.go
//go:build ignore

package main

import (
	"fmt"
)

type abstractApple interface {
	ShowName()
}
type abstractBanana interface {
	ShowName()
}
type ChinaApple struct {
}

func (capple *ChinaApple) ShowName() {
	fmt.Println("中国苹果")
}

type ChinaBanana struct {
}

func (cbanana *ChinaBanana) ShowName() {
	fmt.Println("中国香蕉")
}

type USAApple struct {
}

func (mapple *USAApple) ShowName() {
	fmt.Println("美国苹果")
}

type USABanana struct {
}

func (mbanana *USABanana) ShowName() {
	fmt.Println("美国香蕉")
}

type AbstractFactory interface {
	CreateApple() abstractApple
	CreateBanana() abstractBanana
}
type ChinaFactory struct {
}

func (cfactory *ChinaFactory) CreateApple() abstractApple {
	return new(ChinaApple)
}
func (cfactory *ChinaFactory) CreateBanana() abstractBanana {
	return new(ChinaBanana)
}

type USAFactory struct {
}

func (cfactory *USAFactory) CreateApple() abstractApple {
	return new(USAApple)
}
func (cfactory *USAFactory) CreateBanana() abstractBanana {
	return new(USABanana)
}
func main() {
	var factory AbstractFactory
	var fruit abstractApple
	factory = new(ChinaFactory)
	fruit = factory.CreateApple()
	fruit.ShowName()
	factory = new(USAFactory)
	fruit = factory.CreateBanana()
	fruit.ShowName()
}
